src/Elastic.Apm.Azure.Storage/AzureBlobStorageDiagnosticListener.cs

```diff
@@ -0,0 +1,205 @@
+	{
+		private readonly ConcurrentDictionary<string, IExecutionSegment> _processingSegments =
+			new ConcurrentDictionary<string, IExecutionSegment>();
+
+		public AzureBlobStorageDiagnosticListener(IApmAgent agent) : base(agent) { }
+
+		public override string Name { get; } = "Azure.Storage.Blobs";
+
+		protected override void HandleOnNext(KeyValuePair<string, object> kv)
+		{
+			Logger.Trace()?.Log("Called with key: `{DiagnosticEventKey}'", kv.Key);
+
+			if (string.IsNullOrEmpty(kv.Key))
+			{
+				Logger.Trace()?.Log($"Key is {(kv.Key == null ? "null" : "an empty string")} - exiting");
+				return;
+			}
+
+			switch (kv.Key)
+			{
+				case "BlobContainerClient.Create.Start":
+				case "PageBlobClient.Create.Start":
+					OnStart(kv, "Create");
+					break;
+				case "BlobContainerClient.Delete.Start":
+				case "BlobBaseClient.Delete.Start":
+					OnStart(kv, "Delete");
+					break;
+				case "BlobContainerClient.GetBlobs.Start":
+					OnStart(kv, "GetBlobs");
+					break;
+				case "BlockBlobClient.Upload.Start":
+				case "PageBlobClient.UploadPages.Start":
+					OnStart(kv, "Upload");
+					break;
+				case "BlobBaseClient.Download.Start":
+				case "BlobBaseClient.DownloadContent.Start":
+				case "BlobBaseClient.DownloadStreaming.Start":
+					OnStart(kv, "Download");
+					break;
+				case "BlobBaseClient.StartCopyFromUri.Start":
+					OnStart(kv, "CopyFromUri");
+					break;
+				case "BlobContainerClient.Create.Stop":
+				case "BlobContainerClient.Delete.Stop":
+				case "BlobBaseClient.Delete.Stop":
+				case "PageBlobClient.Create.Stop":
+				case "BlockBlobClient.Upload.Stop":
+				case "BlobBaseClient.Download.Stop":
+				case "BlobBaseClient.DownloadContent.Stop":
+				case "BlobBaseClient.DownloadStreaming.Stop":
+				case "PageBlobClient.UploadPages.Stop":
+				case "BlobContainerClient.GetBlobs.Stop":
+				case "BlobBaseClient.StartCopyFromUri.Stop":
+					OnStop();
+					break;
+				case "BlobContainerClient.Create.Exception":
+				case "BlobContainerClient.Delete.Exception":
+				case "BlobBaseClient.Delete.Exception":
+				case "PageBlobClient.Create.Exception":
+				case "BlockBlobClient.Upload.Exception":
+				case "BlobBaseClient.Download.Exception":
+				case "BlobBaseClient.DownloadContent.Exception":
+				case "BlobBaseClient.DownloadStreaming.Exception":
+				case "PageBlobClient.UploadPages.Exception":
+				case "BlobContainerClient.GetBlobs.Exception":
+				case "BlobBaseClient.StartCopyFromUri.Exception":
+					OnException(kv);
+					break;
+				default:
+					Logger.Trace()?.Log("`{DiagnosticEventKey}' key is not a traced diagnostic event", kv.Key);
+					break;
+			}
+		}
+
+		private void OnStart(KeyValuePair<string, object> kv, string action)
+		{
+			var currentSegment = ApmAgent.GetCurrentExecutionSegment();
+			if (currentSegment is null)
+			{
+				Logger.Trace()?.Log("No current transaction or span - exiting");
+				return;
+			}
+
+			if (!(kv.Value is Activity activity))

```

Comment:
`Activity.Id` according to the API could be `null` - it's very hard for me to imagine that, but to be on the safe side I think null-checking for `activity.Id` would be better, something like this:

```suggestion
			if (!(kv.Value is Activity activity) || string.IsNullOrEmpty(activity.Id))

```

We have the same in the `Elastic.Apm.Azure.ServiceBus` and at other places here (also in the `OnStop` and `OnException` methods).

Other nice thing about null-checking these is that it'll make Rider code analysis happy (currently it shows this as warning for me).

Reply:
[An `Activity` is assigned an id when it's started](https://github.com/dotnet/runtime/blob/6578f257e3be2e2144a65769706e981961f0130c/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs#L623-L670) and `Activity.Current` validates that the value passed to the setter is either a `null` value or a started `Activity` that hasn't finished [[1]](https://github.com/dotnet/runtime/blob/6578f257e3be2e2144a65769706e981961f0130c/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs#L1233-L1242) [[2]](https://github.com/dotnet/runtime/blob/79ae74f5ca5c8a6fe3a48935e85bd7374959c570/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.Current.net46.cs#L16-L26). 

Where the `Activity` is retrieved from the diagnostic event, I think we're safe to assume that the library has started the activity, and where the `Activity` is retrieved from `Activity.Current`, I think we can also assume that this is OK because it will either be the activity from the diagnostic event, or an activity started by the agent in case of a transaction.